src/Prismic/SearchForm.php

```diff
@@ -222,99 +246,125 @@ public function submit()
+        }
+        $this->assertValidQueryParameters($params);
+        if (count($params) === 1 && is_string($first)) {
+            return $this->set("q", $first);
         }
-        $query = "[" . join("", array_map(function($predicate) { return $predicate->q(); }, $predicates)) . "]";
+        $query = "[" . implode("", array_map(function ($predicate) {
+            /** @var Predicate $predicate */
+            return $predicate->q();
+        }, $params)) . "]";
         return $this->set("q", $query);
     }
 
+    /**
+     * Assert that the parameters used for a query contain either a single string, or an array of Predicates
+     * @param array $params
+     * @throws Exception\InvalidArgumentException
+     */
+    private function assertValidQueryParameters(array $params) : void
+    {
+        if (count($params) === 1 && is_string(current($params))) {
+            return;
+        }
+        foreach ($params as $param) {
+            if (! $param instanceof Predicate) {
+                throw new Exception\InvalidArgumentException(
+                    'Query parameters should consist of a single string or multiple Predicate instances'
+                );
+            }
+        }
+    }
+
     /**
      * Get the URL for this form
-     *
-     * @return string the URL
      */
-    public function url()
+    public function url() : string
     {
         $url = $this->form->getAction() . '?' . http_build_query($this->data);
+        /**
+         * This expression removes integer array keys,
+         * i.e. ?q[0]=Whatever&q[1]=OtherThing becomes ?q=Whatever&q=OtherThing
+         */
         $url = preg_replace('/%5B(?:[0-9]|[1-9][0-9]+)%5D=/', '=', $url);
         return $url;
     }
 
     /**
      * Checks if the results for this form are already cached
-     *
-     * @return boolean true if the results for this form are fresh in the cache, false otherwise
      */
-    public function isCached()
+    public function isCached() : bool
     {
-        return $this->api->getCache()->has($this->url());
+        return $this->cache->has($this->url());
     }
 
     /**
      * Performs the actual submit call, without the unmarshalling.
      *
-     * \throws RuntimeException if the Form type is not supported
+     * @throws Exception\RuntimeException if the Form type is not supported or the Response body is invalid
+     * @throws Exception\RequestFailureException if something went wrong retrieving data from the API
      *
-     * @return the raw (unparsed) response.
+     * @return stdClass Unserialized JSON Response
      */
-    private function submit_raw()
+    private function submitRaw() : stdClass
     {
-        if ($this->form->getMethod() == 'GET' &&
-            $this->form->getEnctype() == 'application/x-www-form-urlencoded' &&
-            $this->form->getAction()
+        if ($this->form->getMethod() !== 'GET' ||
+            $this->form->getEnctype() !== 'application/x-www-form-urlencoded' ||
```

Comment:
why change this logic?

Reply:
It's more readable this way around IMO, It reads more like, make sure we have a valid form before attempting anything else… it puts the exception nearer to the reason for it being thrown.

Reply:
seems fair, good for me